### PR TITLE
Feature/file upload

### PR DESCRIPTION
**Pull request by abczzz13**
## What?
Added the file upload functionality for uploading images to AWS S3.

## Why?
This way, admins can upload company logos on their own and we do not depend on logos hosted on other websites where we do not have any control over.

## How?
By using the boto3 libary, a function could be created to upload a file to AWS S3. Then added the API endpoint /upload and included validation to upload an image with a POST request.

## Testing?
Tests are included for testing the good situation and the bad situations to check if responses are as expected. All tests pass.

## Anything else?
Make sure the valid AWS credentials are present as environment variables in the different deployed environments (development, staging, production). Created a new user on AWS with only getObject and putObject access to the fndr bucket on S3.

Also implemented a fix for incorrect links in the pagination meta information when using query parameters. This is used when a GET request is made to /api/v1/companies. (Yes, this should have been made in a different PR)

### CODE REVIEW

File: app/api/companies.py [hunk 1]
```diff
@@ -278,3 +285,34 @@ def delete_company(id):
...
+
+    output = {}
+    output['url'] = str(upload_file_to_s3(img, Config.S3_BUCKET_NAME))
+
+    response = jsonify(output)
+    response.status_code = 201
+
+    return response
+
+    # TODO: We could also think about nameing the file ourselves...
```
Comment (Reinoptland) [inline app/api/companies.py:315]: This seems like a good idea 👍

But then you'd need the rest of the company data (the id and the name etc)

File: app/main/routes.py [hunk 2]
```diff
@@ -35,3 +35,8 @@ def user(username):
 
     user = Users.query.filter_by(username=username).first_or_404()
     return render_template('user.html', title='User Profile', user=user, form=form)
+
+
+@bp.route('/upload')
```
Comment (Reinoptland) [inline app/main/routes.py:40]: I am guessing this is for testing purposes? 🤔

File: app/utils.py [hunk 3]
```diff
@@ -51,3 +80,27 @@ def get_coordinates(city: str) -> dict:
 
         time.sleep(current_delay)
         current_delay *= 2
+
+
+def validate_image(stream):
```
Comment (Reinoptland) [inline app/utils.py:85]: Request validation is always a good idea and checking that you have the right type of file is great first step. 

Additionally you could check for a maximum size, the right aspect ratio (in order to not break the frontend layout). 

It really depends on who you're going to expose a file upload to. If it's the average user it's always good to assume the worst 😂

File: config.py [hunk 4]
```diff
@@ -76,10 +76,17 @@ class Config():
 
     LOG_TO_STDOUT = os.environ.get('LOG_TO_STDOUT')
 
+    MAX_CONTENT_LENGTH = 1024 * 1024
```
Comment (Reinoptland) [inline config.py:79]: Ah! So here is a config that ensures you don't upload something massive 👍

If you break this rule, what kind of error do you get? It's good to test this as you want to display that in the frontend

**Review by Reinoptland — COMMENTED** [review]
Great! 

- You used S3 with the configs in .env 
- There are test cases with happy path and unhappy paths
- You validate a number of things from the incoming file

Consider adding a test case for files that are too large, alternatively you could process the files beyond a certain size to something more manageable

**Review by Reinoptland — APPROVED** [review]

File: app/main/routes.py (hunk 2 above)
Comment (abczzz13) [reply app/main/routes.py:40]: Yes, this was indeed for testing purposes. This whole part should be removed with the Refactor PR.

File: config.py (hunk 4 above)
Comment (abczzz13) [reply config.py:79]: Good point! I will do some tests with this.

File: app/api/companies.py (hunk 1 above)
Comment (abczzz13) [reply app/api/companies.py:315]: I will include this in the issue #22, while also including Pillow to automatically resize the uploaded images.

File: config.py (hunk 4 above)
Comment (abczzz13) [reply config.py:79]: Additionally, by using automatic resizing with Pillow, we can prevent this from happening at all. See issue #22

**Pull request merged**